### DISCUSSION TITLE
RdKit Protein Residues from PDB File

### DISCUSSION THREAD

**Discussion by chezkro70** [rdkit/rdkit#6007] [Q&A] [answered]
Hello

I previously used RdKit to deal with small organic compounds (SDF, SMILES) and to compute various phys-chem properties

However, is there any way to deal with protein-ligand complexes?

I noticed some functions to read PDB files ( Chem.MolFromPDBFile() ) and to add missing hydrogen atoms ... but I did not find a way to:
- loop over protein residues, or
- extract, for instance, all complete residues around a given ligand or residue ID (e.g., around 6A)

Is it possible using python Rdkit scripts? If yes, can you give me some hints?

Thank you for appreciated help!

Alex

**Comment by ptosco** [comment, marked as answer]
@chezkro70 Residue information is stored on a per-atom basis on molecules parsed with `Chem.MolFromPDBFile()` and `Chem.MolFromPDBBlock()`, so, yes, this is possible.
If you look into [BioPython](https://biopython.org/) you will find more readily available functionality to carry out this sort of tasks on macromolecular complexes.
That being said, the snippet below exemplifies how to extract all complete residues within a threshold from a specific residue using RDKit:
```
import requests
from rdkit import Chem
import numpy as np

# Load 1OIT PDB complex
url = "https://files.rcsb.org/download/1OIT.pdb"
r = requests.get(url, allow_redirects=True)
assert r.status_code == 200
pdb_complex = Chem.MolFromPDBBlock(r.content.decode("utf-8"), removeHs=False)

# Extract HDT ligand
hdt = list(filter(lambda frag: frag.GetNumAtoms()
    and frag.GetAtomWithIdx(0).GetPDBResidueInfo().GetResidueName() == "HDT",
    Chem.GetMolFrags(pdb_complex, asMols=True)))
assert len(hdt) == 1
hdt = hdt.pop(0)

def closer_than(atom_pos, residue_positions, threshold):
    """Return True if any coord tuple in residue_positions is closer than threshold to atom_pos."""
    return any(np.linalg.norm(atom_pos - pos) < threshold for pos in residue_positions)

# Get coord tuple arrays for HDT residue and whole PDB complex
residue_positions = hdt.GetConformer().GetPositions()
complex_positions = pdb_complex.GetConformer().GetPositions()

THRESHOLD = 6

# Collect residue numbers within threshold
residues_within_threshold = set()
for i, complex_pos in enumerate(complex_positions):
    resnum = pdb_complex.GetAtomWithIdx(i).GetPDBResidueInfo().GetResidueNumber()
    if resnum not in residues_within_threshold and closer_than(complex_pos, residue_positions, THRESHOLD):
        residues_within_threshold.add(resnum)

# Remove all atoms beyond threshold
with Chem.RWMol(pdb_complex) as pdb_residues_within_threshold:
    for a in pdb_residues_within_threshold.GetAtoms():
        if a.GetPDBResidueInfo().GetResidueNumber() not in residues_within_threshold:
            pdb_residues_within_threshold.RemoveAtom(a.GetIdx())

# Save residues within threshold
Chem.MolToPDBFile(pdb_residues_within_threshold, "residues_within_threshold.pdb")
```
![image](https://user-images.githubusercontent.com/5244385/214707852-aad8cc30-e7c8-4a67-8ffb-195f11e0a812.png)

**Reply by chezkro70** [reply]
@ptosco thank you for your quick and interesting help. I will try your solution. I will also take a look at BioPython.

**Comment by mwojcikowski** [comment]
I can also recommend a function `ExtractPocketAndLigand `  that I wrote in ODDT some time ago that uses RDKit directly https://github.com/oddt/oddt/blob/master/oddt/toolkits/extras/rdkit/fixer.py#L166-L272

**Reply by chezkro70** [reply]
Thank you for this alternative solution
